phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexIT.java

```diff
@@ -259,7 +279,150 @@ private void assertIndexMutations(Connection conn) throws SQLException {
+                try {
+                    upsertRows(conn, fullTableName, numRows);
+                } catch (CommitException ex) {
+                    isWriteOnDisabledIndexFailed = true;
+                }
+                assertEquals(true, isWriteOnDisabledIndexFailed);
+                PIndexState indexState = TestUtil.getIndexState(conn, fullIndexName);
+                assertEquals(PIndexState.ACTIVE, indexState);
+            }
+        }
+    }
+
+    @Test
+    public void testGlobalImmutableIndexDelete() throws Exception {
+        if (localIndex || transactionProvider != null) {
+            return;
+        }
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String tableName = "TBL_" + generateUniqueName();
+        String indexName = "IND_" + generateUniqueName();
+        String fullTableName = SchemaUtil.getTableName(TestUtil.DEFAULT_SCHEMA_NAME, tableName);
+        String fullIndexName = SchemaUtil.getTableName(TestUtil.DEFAULT_SCHEMA_NAME, indexName);
+        TABLE_NAME = fullTableName;
+        try (Connection conn = DriverManager.getConnection(getUrl(), props);
+                Admin admin = conn.unwrap(PhoenixConnection.class).getQueryServices().getAdmin();) {
+            conn.setAutoCommit(true);
+            int numRows = 2;
+            createAndPopulateTableAndIndexForConsistentIndex(conn, fullTableName, fullIndexName, numRows);
+
+            String dml = "DELETE from " + fullTableName + " WHERE varchar_pk='varchar1'";
+            conn.createStatement().execute(dml);
+            conn.commit();
+            ResultSet rs;
+            rs = conn.createStatement().executeQuery("SELECT /*+ NO_INDEX */ COUNT(*) FROM " + TABLE_NAME);
+            assertTrue(rs.next());
+            assertEquals(numRows - 1, rs.getInt(1));
+            rs = conn.createStatement().executeQuery("SELECT COUNT(*) FROM " + fullIndexName);
+            assertTrue(rs.next());
+            assertEquals(numRows - 1, rs.getInt(1));
+
+            // Force delete to fail (data removed but operation failed) on data table and check index table row remains as unverified
+            TestUtil.addCoprocessor(conn, fullTableName, DeleteFailingRegionObserver.class);
+            dml = "DELETE from " + fullTableName + " WHERE varchar_pk='varchar2'";
+            boolean isDeleteFailed = false;
+            try {
+                conn.createStatement().execute(dml);
+            } catch (Exception ex) {
+                isDeleteFailed = true;
+            }
+            assertEquals(true, isDeleteFailed);
+            TestUtil.removeCoprocessor(conn, fullTableName, DeleteFailingRegionObserver.class);
+            assertEquals(numRows - 1, getRawRowCount(conn.unwrap(PhoenixConnection.class).getQueryServices()
+                    .getTable(Bytes.toBytes(fullIndexName)), false));
+            assertEquals(true, verifyRowsForEmptyColValue(conn, fullIndexName, IndexRegionObserver.UNVERIFIED_BYTES));
```

Comment:
Do we expect that all the rows are unverified here? It seems only one of the rows should be unverified.

Reply:
We have 1 row here. 1 of them is deleted on line 359

Reply:
I missed that you set numRows.